### DISCUSSION TITLE
Infinity Loop / Players Average

### DISCUSSION THREAD

**Discussion by fimoefive** [nss-evening-cohort-14/evening-client-side#335] [Capstones!] [answered]
Martin

**issue:**
I've successfully wrote an infinity for loop which keeps freezing the DOM.

I'm trying to import the holes data into games.js file and loop through all the holes[I], add, and divide by hole += Number(holeInfo[I]) / holes.length;
For Players Average.

Mainly having trouble grabbing holes info within games.js file in the views folder, so I loop through all the holes the player has played and divide by the length of holes.

**What I have tried:**
Wrote a for loop and a .reduce average Method functions

  function playerAvg(gameInfo) {
    // console.warn(gameInfo);
    const gamesAverage = getHoles().length;
    console.warn(gamesAverage);
    let average = 0;
    for (let i = 0; i < gamesAverage; i) {
      average = Number(gameInfo) || 0;
    }
    return average;
  }

  // const totalAvg = getHoles(holes.uid).then((gamesArray) => setHoles(gamesArray)).reduce(
  //   (previousScore, currentScore) => previousScore + currentScore,
  //   0
  // );
  // console.warn(totalAvg);

**The current Firebase rules to:**
            "games": {
        ".indexOn": ["uid", "gameFirebaseKey"]
      }, 
            "holes": {
        ".indexOn": ["uid","hole1", "hole2", "hole3", "hole4", "hole5", "hole6", "hole7", "hole8", "hole9", "avg", "total", ".value"]
      }, 
                    "user": {
        ".indexOn": ["uid"]
      }, 
  }

**Google links:**
https://stackoverflow.com/questions/43260689/adding-and-indexon-value-into-firebase-rules
https://jrsinclair.com/articles/2019/five-ways-to-average-with-js-reduce/
https://dev.to/yogesnsamy/how-to-use-the-reduce-method-in-javascript-and-react-5dhl

**Links to code:**
https://github.com/fimoefive/Golf-ScoreCard-App/blob/mjs-Avgs/src/views/Games.js

**Comment by TrinityChristiana** [comment]
`getHoles` is a promise. You can't get `.length` of a promise. You'll have to wait for the response of `getHoles`, then get the length of its response.

Also, remember to add the number from `gameInfo` to the `average` instead of resetting the value of `average`

**Reply by fimoefive** [reply]
Do I need holes, setHoles useState in Games.js file?

**Reply by TrinityChristiana** [reply]
I don't think so.

**Reply by fimoefive** [reply]
I'm having trouble getting the response of getHoles() promise 
 function playerAvg(gameInfo, resolve) {
    const gamesAverage = getHoles().then((response) => resolve(Object.values(response.data))).length;
    // console.warn(gamesAverage);
    let average = 0;
    for (let i = 0; i < gamesAverage; i) {
      average += Number(gameInfo);
    }
    console.warn(average);
    return average;
  }

**Comment by TrinityChristiana** [comment, marked as answer]
@fimoefive Will make the talked about updates and will update the ticket with the solution.

**Comment by fimoefive** [comment]
I declared a playerAverage setPlayerAverage = useState
Had to switch to a .forEach Method loop in the useEffect within the useEffect 

  useEffect(() => {
    getHoles(user.uid).then((holesArray) => {
      let average = 0;
      holesArray.forEach((holes) => {
        average += Number(holes.hole1);
        average += Number(holes.hole2);
        average += Number(holes.hole3);
        average += Number(holes.hole4);
        average += Number(holes.hole5);
        average += Number(holes.hole6);
        average += Number(holes.hole7);
        average += Number(holes.hole8);
        average += Number(holes.hole9);
      });
setPlayerAverage(average / holesArray.length);
    });
  }, []);
you have to write a // for (let i = 0; i < holesArray.length; i = +1) { 
like this in es lint
